dipy/reconst/multi_voxel.py

```diff
@@ -5,6 +5,224 @@
+
+
+_numpy_to_ctypes = dict(zip(_ctypes_to_numpy.values(),
+                            _ctypes_to_numpy.keys()))
+
+global shared_arr
+
+
+def shm_as_ndarray(mp_array, shape=None):
+    """
+    Given a multiprocessing.Array, returns an ndarray pointing to
+    the same data.
+
+    Parameters
+    ----------
+    mp_array : array
+        multiprocessing.Array that you want to convert.
+    shape : tuple
+        tuple of array dimensions (default: None)
+
+    Returns
+    --------
+    numpy_array : ndarray
+        ndarray pointing to the same data
+    """
+
+    # support SynchronizedArray:
+    if not hasattr(mp_array, '_type_'):
+        mp_array = mp_array.get_obj()
+
+    dtype = _ctypes_to_numpy[mp_array._type_]
+    result = np.frombuffer(mp_array, dtype)
+
+    if shape is not None:
+        result = result.reshape(shape)
+
+    return np.asarray(result)
+
+
+def ndarray_to_shm(arr, lock=False):
+    """
+    Generate an 1D multiprocessing.Array containing the data from
+    the passed ndarray.  The data will be *copied* into shared
+    memory.
+
+    Parameters
+    ----------
+    arr : ndarray
+        numpy ndarray that you want to convert.
+    lock : boolean
+        controls the access to the shared array. When your shared
+        array has a read only access, you do not need lock. Otherwise,
+        any writing access need to activate the lock to be
+        process-safe(default: False)
+
+    Returns
+    --------
+    shm : multiprocessing.Array
+        copied shared array
+    """
+
+    array1d = arr.ravel(order='A')
+
+    try:
+        c_type = _numpy_to_ctypes[array1d.dtype]
+    except KeyError:
+        c_type = _numpy_to_ctypes[np.dtype(array1d.dtype)]
+
+    result = Array(c_type, array1d.size, lock=lock)
+    shm_as_ndarray(result)[:] = array1d
+
+    return result
+
+
+def parallel_fit_worker(arguments):
+    """
+    Each pool process calls this worker.
+    Fit model on chunks
+
+    Parameters
+    -----------
+    arguments: tuple
+        tuple should contains a model and
+        list of indexes
+
+    Returns
+    --------
+    result: list of tuple
+        return a list of tuple(voxel index, model fitted instance)
+    """
+    model, chunks = arguments
+    return [(idx, model.fit(shared_arr[idx]))
+            for idx in chunks]
+
+
+def _init_parallel_fit_worker(arr_to_populate, shape):
+    """
+    Each pool process calls this initializer.
+    Load the array to be populated into that process's global namespace
+
+    Parameters
+    -----------
+    arr_to_populate: multiprocessing.Array
+        shared array
+    shape: tuple
+        tuple of array dimensions
+    """
+    global shared_arr
+    with warnings.catch_warnings():
+        warnings.simplefilter('ignore', RuntimeWarning)
+        # shared_arr = np.ctypeslib.as_array(arr_to_populate)
+        # shared_arr = shared_arr.reshape(shape)
+        shared_arr = shm_as_ndarray(arr_to_populate, shape)
+
```

Comment:
I would love to have some unit tests for the functions above. If I understand correctly, right now they are tested only through the decorator.

Reply:
That's true, I will add them.

BTW, I still wonder if these functions (`shm_as_array` and `ndarray_to_shm`) are in the right place because they can be used for other purposes like `peaks_from_model`

Reply:
Maybe in a new `dipy.core.parallel` module?